### DISCUSSION TITLE
Restart sampler custom advanced?

### DISCUSSION THREAD

**Discussion by blepping** [ssitu/ComfyUI_restart_sampling#17] [Ideas]
currently there isn't a way to use restart with the new `Guider` and `Noise` stuff. i haven't looked at it too closely yet so i'm not sure how big of a change it will be, but would you be interested in a pull adding that feature?

**Comment by ssitu** [comment]
Are those new additions to comfyui?

**Comment by blepping** [comment]
> Are those new additions to comfyui?

yes, that's correct. the `SamplerCustomAdvanced` node was added on the 3rd i think, so pretty recent. a bit more than a week ago:

![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/d873ae78-167c-4be5-ac70-3cd36abf07b8)

aside from the guider replacing conditioning and the noise input addition it works like the `SamplerCustom` node.

**Comment by ssitu** [comment]
Looks like it is to support more advanced sampling schemes. The dual cfg guider sounds interesting. I think it'll be good to support the sampler custom advanced node.

I wonder if we can just have a schedule-related node and a guider node to support this, where the schedule node inserts the restarts and the guider node adds noise back when the restarts occur. I haven't looked into the new code too much, so it is just a thought. I forgot how hard it was to follow the sampling code.

**Comment by blepping** [comment]
i looked at it more closely the other day. for reference the class ComfyUI defines: https://github.com/comfyanonymous/ComfyUI/blob/58812ab8ca601cc2dd9dbe64c1f3ffd4929fd0ca/comfy/samplers.py#L592

from that, it looks like making a custom advanced version sampler wouldn't really be too useful (and possibly would be difficult/impossible). also making a restart guider that takes stuff like the restart segments/schedulers could work but might be kind of pointless because you couldn't use another guider. so it's kind of back to doing what you already can do, except for the noise input i guess.

what might be possible is to make a restart guider that takes another guider. then you could actually do stuff with it that's not possible currently. doing it that way there would be a restart guider that takes another guider and also a node that takes the restart params like segments/schedules creates the restart sigmas.

(maybe this should be moved to a discussion)

**Reply by ssitu** [reply]
> (maybe this should be moved to a discussion)

Done!

> what might be possible is to make a restart guider that takes another guider. then you could actually do stuff with it that's not possible currently. doing it that way there would be a restart guider that takes another guider and also a node that takes the restart params like segments/schedules creates the restart sigmas.

It seems to me that the guider has access to model output, so that could be where the renoising happens during a restart. Then the sigmas input to the custom sampler node can come from a node that just adds restarts to a given sigma schedule.

I think the goal would be to get the DualCFG working with Restarts, and any other guider that may exist.

**Reply by blepping** [reply]
i've been messing with it, and i don't think it needs to use the guider stuff at all. it can just be a `SAMPLER`: https://github.com/ssitu/ComfyUI_restart_sampling/pull/18

that pull is obviously far from ready. there are also some issues with the approach i'm using. what i'd like to do is have a node to generate the restart sigmas but it seems impossible to recover where segments begin/end. the approach i took is having restart sigmas be negative, but there's also an issue there: the start/end sigmas in a restart segment aren't necessarily the same as s_min/s_max and the restart noise should be generated with that (but since i only have access to the segment i use that instead).

it might make more sense to have the restart scheduler node just generate a restart-compatible set of sigmas and then have the restart sampler node be what takes the segments/restart schedule and inserts the segment sigmas.

right now it looks like:
![image](https://github.com/ssitu/ComfyUI_restart_sampling/assets/157360029/51493410-b93b-4717-a106-1cb4620a4b8e)

so this would move segments and restart_scheduler to the `RestartScheduler` node.

any thoughts on what approach to use, or also if there is a way to recover segment start/end points _and_ smin/smax given only the list of sigmas?

**Reply by blepping** [reply]
i _think_ i might have figured it out (at least if my assumptions are correct). s_min should be the last sigma in the previous non-restart segment. s_max should just be the first sigma in the restart segment. also the first sigma after a restart segment will be the last sigma from the previous non-restart segment. the restart segment might end with that sigma also, but since sigmas won't be repeated it _should_ be possible to figure out where segments begin/end and the correct s_min/s_max for noising the restart segments.

i will do more testing tomorrow. please correct me if those assumptions are wrong

**Reply by ssitu** [reply]
> but there's also an issue there: the start/end sigmas in a restart segment aren't necessarily the same as s_min/s_max and the restart noise should be generated with that

When would they be different? Here's a dump for a schedule with restarts:
```
** Dumping restart sampling plan (total steps 14):
[    ] Step   1..6  : tensor([14.6146,  9.1029,  5.4784,  3.1686,  1.7494,  0.9141,  0.4469])
[R  1] Step   7..8  : tensor([0.8319, 0.6140, 0.4469])
[R  2] Step   9..10 : tensor([0.8319, 0.6140, 0.4469])
[    ] Step  11..14 : tensor([0.4469, 0.2014, 0.0819, 0.0292, 0.0000])
```
Wouldn't s_max be 0.8319 and s_min be 0.4469 for the two restart segments in this case? Maybe I'm forgetting how it works.

**Reply by blepping** [reply]
might be an indication something else is wrong but i can get different results in some cases:

```plaintext
36 steps, normal, normal, default:
[R  1] Step  33..34 : 0.5908 (0.59), 0.4493, 0.3121 (0.3112)
[R  1] Step  38..39 : 0.3006 (0.3), 0.2004, 0.02917 (same)

36 steps, ddim_uniform, ddim_uniform, default:
[R  1] Step  33..34 : 0.5908 (0.59), 0.4471, 0.3098 (same)
[R  1] Step  38..39 : 0.3006 (0.3), 0.1991, 0.02917 (same)

36 steps, simple_test, simple_test, default:
[R  1] Step  34..35 : 0.5884 (0.59), 0.4924, 0.3977 (0.3052)
[R  1] Step  39..40 : 0.2983 (0.3), 0.2332, 0.158 (0.0 ???)

36 steps, normal, normal, a1111:
[R  1] Step  35..43 : 2.062 (2.061), 1.678, 1.374, 1.128, 0.9233, 0.7485, 0.5949, 0.4548, 0.3195, 0.1673 (0.1659)

36 steps, simple_test, simple_test, a1111:
[R  1] Step  36..44 : 1.937 (1.943), 1.624, 1.363, 1.145, 0.9643, 0.8041, 0.6625, 0.5363, 0.4156, 0.2959 (0.1611)
```

maybe `simple_test` isn't expected to work but there's still 0.3121 vs 0.3112 with the normal schedule. if they're supposed to match exactly, can we just change the thing that builds restart segments to overwrite the first sigma in the segment to s_max and the last one to s_min?

**Reply by ssitu** [reply]
Are the values in the parentheses the ones you get from your PR and the rest are from the current implementation?

**Reply by blepping** [reply]
sorry for not explaining. the values in parenthesis are the t_min/t_max next to the min/max values for the actual segment sigmas to illustrate the difference.

i've made a fair amount of progress on #18, i think at this point it actually is functional though it still needs testing and cleanup.

i also added a self-test function and ran it on all scheduler, restart scheduler, default and a1111 segments from steps 2 to 200 + comparing the recovered plan and it seems to work.

the `simple_test` schedule just... doesn't seem to work though. what is it supposed to do? using it creates an invalid plan a lot of the time. also even with other schedulers it seems possible for restart sigmas to end lower than the the next normal segment. i had to adjust the self test sensitive for that to only checking up to two decimal places.

now that i had you create a discussion, maybe it's time to move again - to the pull! sorry, at that point i wasn't sure how far i was from an actual working implementation and didn't know how much discussion would be required to just figure out the approach to use.

**Reply by ssitu** [reply]
> now that i had you create a discussion, maybe it's time to move again - to the pull! sorry, at that point i wasn't sure how far i was from an actual working implementation and didn't know how much discussion would be required to just figure out the approach to use.

All good. I'll reply to this on your PR.
